File: DiabloCharacterTracker.Server/Services/SkillServices/SkillService.cs
using DiabloCharacterTracker.Server.Data;
using DiabloCharacterTracker.Server.DTOs;
using Microsoft.EntityFrameworkCore;

namespace DiabloCharacterTracker.Server.Services.SkillServices;

public class SkillService : ISkillService
{
    private readonly IDbContextFactory<DiabloDbContext> dbContextFactory;

    public SkillService(IDbContextFactory<DiabloDbContext> dbContextFactory)
    {
        this.dbContextFactory = dbContextFactory;
    }

    public async Task<List<SkillDTO>> GetAllSkills()
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var skills = await context.Skills.ToListAsync();

        if (skills == null)
        {
            return new List<SkillDTO>();
        };

        return skills.Select(x => x.ToDTO()).ToList();
    }

    public async Task<SkillDTO> GetSkillById(int skillId)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var skill = await context.Skills
            .Where(x => x.Id == skillId)
            .FirstOrDefaultAsync();

        if (skill == null)
        {
            return new SkillDTO();
        }

        return skill.ToDTO();
    }

    public async Task<SkillDTO> GetSkillByName(string skillName)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var skill = await context.Skills
            .Where(x => x.SkillName == skillName)
            .FirstOrDefaultAsync();

        if (skill == null)
        {
            return new SkillDTO(); 
        }

        return skill.ToDTO();   
    }

    public async Task<List<SkillDTO>> GetAllSkillsByClassName(string className)
    {
        using var context = await dbContextFactory.CreateDbContextAsync();

        var skills = await context.Skills
            .Where(x => x.ClassName == className)
            .ToListAsync();

        if (skills == null)
        {
            return new List<SkillDTO>();
        }

        return skills.Select(x => x.ToDTO()).ToList();  
    }
}

File: diablocharactertracker.client/src/ApiServices/SkillService.ts
import { Skill } from "../Data/DTOs/Skill";
import axios from "axios";

export const SkillService = {
  GetAllSkills: async () => {
    try {
      const response = await axios.get<Skill[]>(
        `${import.meta.env.VITE_URL}/Skill/getallskills`
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get all skills");
      throw error;
    }
  },
  GetSkillById: async (skillId: number) => {
    try {
      const response = await axios.get<Skill>(
        `${import.meta.env.VITE_URL}/Skill/getskillbyid`,
        {
          params: {
            skillId,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get skill by id");
      throw error;
    }
  },
  GetSkillByName: async (skillName: string) => {
    try {
      const response = await axios.get<Skill>(
        `${import.meta.env.VITE_URL}/Skill/getskillbyname`,
        {
          params: {
            skillName,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get skill by name");
      throw error;
    }
  },
  GetAllSkillsByClassName: async (className: string) => {
    try {
      const response = await axios.get<Skill[]>(
        `${import.meta.env.VITE_URL}/Skill/getallskillsbyclassname`,
        {
          params: {
            className,
          },
        }
      );
      return response.data;
    } catch (error) {
      console.error("Failed to get all skill for class");
      throw error;
    }
  },
};
